### PR TITLE
Resize and EP specific transpose optimization updates

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

- Treat Resize as layout sensitive by default
  - whilst the ONNX spec does not specify a layout, EPs tend to implement only one
  - add second usage in L2 of TransposeOptimizer to plugin the ability to push a Transpose through a Resize assigned to the CPU EP
- Allow EP specific logic for changes the ops considered to be layout sensitive to be plugged in
  - expected usage is for #17200 

### Motivation and Context

Finish simplifying/clarifying transpose optimization and layout transformation that was proposed in #15552. This PR along with #17618 should complete the changes.

### CODE REVIEW

File: onnxruntime/test/optimizer/transpose_optimizer_test.cc [hunk 1]
```diff
@@ -4450,12 +4449,13 @@ TEST(TransposeOptimizerTests, RegressionTest_GitHubIssue12151) {
               testing::ContainerEq(fetches[0].Get<Tensor>().DataAsSpan<float>()));
 }
 
+// These tests uses internal testing EP with static kernels which requires a full build,
+// and the NHWC Conv with requires contrib ops
```
Comment (edgchen1) [inline onnxruntime/test/optimizer/transpose_optimizer_test.cc:4453]:
```suggestion
// and the NHWC Conv which requires contrib ops
```

**Review by edgchen1 — DISMISSED** [review]

**Review by adrianlizarraga — DISMISSED** [review]

File: onnxruntime/core/optimizer/transpose_optimization/ort_transpose_optimization.cc [hunk 2]
```diff
@@ -5,12 +5,35 @@
...
 namespace onnxruntime {
 
+static bool EPAwareHandleResize(HandlerArgs& args) {
+  // Whilst Resize is not technically layout sensitive, execution providers typically implement handling for only one
+  // layout. Due to that, only push a Transpose through a Resize once it is assigned and we know it's being handled
+  // by an EP that supports multiple layouts. Currently that's the CPU and XNNPACK EPs.
+  const auto ep_type = args.node.GetExecutionProviderType();
+  if (ep_type == kCpuExecutionProvider || ep_type == kXnnpackExecutionProvider) {
+    // allow NCHW <-> NHWC for now. not clear any other sort of transpose has a valid usage in a real model
+    int64_t rank_int = gsl::narrow_cast<int64_t>(args.perm.size());
```
Comment (adrianlizarraga) [inline onnxruntime/core/optimizer/transpose_optimization/ort_transpose_optimization.cc:22]: nit: doesn't seem to require a narrow_cast from size_t to int64_t.

**Review by adrianlizarraga — APPROVED** [review]

File: onnxruntime/core/optimizer/transpose_optimization/ort_transpose_optimization.cc (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/optimizer/transpose_optimization/ort_transpose_optimization.cc:22]: IIUC int64_t is signed and size_t isn't, so for 64-bit they have different ranges even though they're the same number of bits and the 'narrowing' here is dealing with values outside that range (which will never happen as the size of the vector will never be negative so the compiler doesn't know that).

Comment (adrianlizarraga) [reply onnxruntime/core/optimizer/transpose_optimization/ort_transpose_optimization.cc:22]: Right, but this `rank_int` is only just compared to the literal `4`, so making it signed for this comparison seems unnecessary (but harmless).

**Pull request merged**